### PR TITLE
Migrations

### PR DESCRIPTION
**Pull request by StijnVrolijk**
This PR adds migrations to our deploy script. This is capistrano 3 conversion of the migration tasks in https://github.com/sumocoders/forkcms_deploy

### CODE REVIEW

File: lib/maintenance/.htaccess [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# All url's will be see the maintenance page (index.html)
```
Comment (tijsverkoyen) [inline lib/maintenance/.htaccess:1]: typo

File: lib/capistrano/tasks/migrations.rake [hunk 2]
```diff
@@ -0,0 +1,169 @@
+namespace :sumo do
+  namespace :migrations do
+    desc 'prepares the server for running fork migrations'
+    task :prepare do
```
Comment (tijsverkoyen) [inline lib/capistrano/tasks/migrations.rake:4]: I would separate this in specific methods, so they have less responsibility

File: lib/capistrano/tasks/migrations.rake [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+          # Fetch all migration directories that aren't executed yet
+          folders.each do |dirname|
+            migrationsToExecute.push(dirname) if executedMigrations.index(dirname) == nil
+          end
+
+          if migrationsToExecute.length > 0
+            # This can take a while and can go wrong. let's show a maintenance page
+            # and make sure we can put back the database
+            invoke 'sumo:migrations:symlink_maintenance'
+            invoke 'sumo:migrations:backup_database'
```
Comment (tijsverkoyen) [inline lib/capistrano/tasks/migrations.rake:75]: Wouldn't it be better, more logical to have a backup-task in the db-namespace?

File: lib/capistrano/tasks/migrations.rake [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+            # run all migrations
+            migrationsToExecute.each do |dirname|
+              migrationpath = "#{release_path}/migrations/#{dirname}"
+              migrationFiles = capture("ls -1 #{migrationpath}").split(/\r?\n/)
+
+              migrationFiles.each do |filename|
+                if filename.index('locale.xml') != nil
+                  execute :php, "#{release_path}/bin/console forkcms:locale:import -f #{migrationpath}/#{filename}"
```
Comment (tijsverkoyen) [inline lib/capistrano/tasks/migrations.rake:84]: This is fork specific, so I would PR the whole change to tijsverkoyen/capistrano-forkcms instead.

File: lib/capistrano/tasks/migrations.rake [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                end
+
+                if filename.index('update.php') != nil
+                  execute :php, "#{migrationpath}/#{filename}"
```
Comment (tijsverkoyen) [inline lib/capistrano/tasks/migrations.rake:88]: I would create separate tasks for locale, php-scripts, SQL-updates

File: lib/capistrano/sumo/defaults.rb [hunk 6]
```diff
@@ -1,14 +1,20 @@
 namespace :deploy do
   # compile and upload the assets
   after :updated, 'sumo:assets:put'
+  after :updated, 'sumo:migrations:prepare'
```
Comment (tijsverkoyen) [inline lib/capistrano/sumo/defaults.rb:4]: I think it would be more logical to invoke this in the namespace itself instead of calling it here.

**Review by tijsverkoyen — CHANGES_REQUESTED** [review]

**Comment by StijnVrolijk** [comment]
Created some PR's on https://github.com/tijsverkoyen/capistrano-forkcms/

**Pull request closed without merging**